File: backend/src/main/kotlin/dev/reviewengine/domain/Scale.kt
package dev.reviewengine.domain

import java.math.BigDecimal
import java.math.MathContext

/**
 * An exact decimal score scale. Scores are persisted as integer ticks, never floating point.
 */
class Scale private constructor(
    val minValue: BigDecimal,
    val maxValue: BigDecimal,
    val stepValue: BigDecimal,
    val maxTick: Long,
) {
    companion object {
        fun of(
            minValue: String = "0",
            maxValue: String,
            stepValue: String,
        ): Scale {
            val min = minValue.toExactDecimal("minValue")
            val max = maxValue.toExactDecimal("maxValue")
            val step = stepValue.toExactDecimal("stepValue")

            if (min >= max) {
                domainFailure(
                    DomainErrorCode.INVALID_SCALE,
                    "Scale minimum must be less than its maximum",
                    "minValue" to minValue,
                    "maxValue" to maxValue,
                )
            }
            if (step <= BigDecimal.ZERO) {
                domainFailure(
                    DomainErrorCode.INVALID_SCALE,
                    "Scale step must be greater than zero",
                    "stepValue" to stepValue,
                )
            }

            val ticks = (max - min).divideAndRemainder(step)
            if (ticks[1].compareTo(BigDecimal.ZERO) != 0) {
                domainFailure(
                    DomainErrorCode.INVALID_SCALE,
                    "Scale range must be exactly divisible by its step",
                    "minValue" to minValue,
                    "maxValue" to maxValue,
                    "stepValue" to stepValue,
                )
            }

            val maxTick = try {
                ticks[0].longValueExact()
            } catch (exception: ArithmeticException) {
                throw DomainException(
                    DomainErrorCode.INVALID_SCALE,
                    "Scale contains too many ticks",
                    mapOf("maxTick" to ticks[0].toPlainString()),
                    exception,
                )
            }
            return Scale(min, max, step, maxTick)
        }

        private fun String.toExactDecimal(field: String): BigDecimal = try {
            BigDecimal(this)
        } catch (exception: NumberFormatException) {
            throw DomainException(
                DomainErrorCode.INVALID_DECIMAL,
                "$field must be a decimal string",
                mapOf(field to this),
                exception,
            )
        }
    }

    fun isValidTick(tickIndex: Long): Boolean = tickIndex in 0..maxTick

    fun requireTick(tickIndex: Long) {
        if (!isValidTick(tickIndex)) {
            domainFailure(
                DomainErrorCode.INVALID_TICK_INDEX,
                "Tick index is outside the scale",
                "tickIndex" to tickIndex,
                "minTick" to 0,
                "maxTick" to maxTick,
            )
        }
    }

    fun display(tickIndex: Long): BigDecimal {
        requireTick(tickIndex)
        return minValue + stepValue.multiply(BigDecimal.valueOf(tickIndex))
    }

    fun displayString(tickIndex: Long): String = display(tickIndex).toCanonicalString()

    fun normalized(tickIndex: Long): BigDecimal {
        requireTick(tickIndex)
        return BigDecimal.valueOf(tickIndex).divide(BigDecimal.valueOf(maxTick), MathContext.DECIMAL128)
    }

    fun normalizedString(tickIndex: Long): String = normalized(tickIndex).toCanonicalString()

    fun minValueString(): String = minValue.toCanonicalString()

    fun maxValueString(): String = maxValue.toCanonicalString()

    fun stepValueString(): String = stepValue.toCanonicalString()

    override fun equals(other: Any?): Boolean = other is Scale &&
        minValue.compareTo(other.minValue) == 0 &&
        maxValue.compareTo(other.maxValue) == 0 &&
        stepValue.compareTo(other.stepValue) == 0

    override fun hashCode(): Int {
        var result = minValue.stripTrailingZeros().hashCode()
        result = 31 * result + maxValue.stripTrailingZeros().hashCode()
        result = 31 * result + stepValue.stripTrailingZeros().hashCode()
        return result
    }

    override fun toString(): String = "Scale(min=${minValueString()}, max=${maxValueString()}, step=${stepValueString()})"
}

private fun BigDecimal.toCanonicalString(): String = stripTrailingZeros().toPlainString()

File: worker/src/scale.ts
import Decimal from 'decimal.js';
import { HttpError, requireInteger } from './http';

Decimal.set({ precision: 40, rounding: Decimal.ROUND_HALF_EVEN, toExpNeg: -1e9, toExpPos: 1e9 });

function decimal(value: unknown, field: string): Decimal {
  if (typeof value !== 'string') {
    throw new HttpError(422, 'INVALID_DECIMAL', `${field} must be a decimal string`, { [field]: String(value ?? '') });
  }
  try {
    return new Decimal(value);
  } catch {
    throw new HttpError(422, 'INVALID_DECIMAL', `${field} must be a decimal string`, { [field]: value });
  }
}

function canonical(value: Decimal): string {
  const result = value.toFixed();
  if (!result.includes('.')) return result === '-0' ? '0' : result;
  const trimmed = result.replace(/(?:\.0+|(?:(\.\d*?)0+))$/, '$1');
  return trimmed === '-0' ? '0' : trimmed;
}

export class Scale {
  readonly min: Decimal;
  readonly max: Decimal;
  readonly step: Decimal;
  readonly maxTick: number;

  constructor(minValue: unknown, maxValue: unknown, stepValue: unknown) {
    this.min = decimal(minValue, 'minValue');
    this.max = decimal(maxValue, 'maxValue');
    this.step = decimal(stepValue, 'stepValue');
    if (!this.min.lt(this.max)) {
      throw new HttpError(422, 'INVALID_SCALE', 'Scale minimum must be less than its maximum');
    }
    if (!this.step.gt(0)) {
      throw new HttpError(422, 'INVALID_SCALE', 'Scale step must be greater than zero');
    }
    const ticks = this.max.minus(this.min).div(this.step);
    if (!ticks.isInteger() || ticks.lt(0) || ticks.gt(Number.MAX_SAFE_INTEGER)) {
      throw new HttpError(422, 'INVALID_SCALE', ticks.isInteger() ? 'Scale contains too many ticks' : 'Scale range must be exactly divisible by its step');
    }
    this.maxTick = ticks.toNumber();
  }

  requireTick(value: unknown): number {
    const tick = requireInteger(value, 'tickIndex', 0);
    if (tick > this.maxTick) {
      throw new HttpError(422, 'INVALID_TICK_INDEX', 'Tick index is outside the scale', {
        tickIndex: String(tick), minTick: '0', maxTick: String(this.maxTick),
      });
    }
    return tick;
  }

  display(tick: number): string {
    return canonical(this.min.plus(this.step.times(tick)));
  }

  normalized(tick: number): number {
    return new Decimal(tick).div(this.maxTick).toNumber();
  }

  minValue(): string { return canonical(this.min); }
  maxValue(): string { return canonical(this.max); }
  stepValue(): string { return canonical(this.step); }
}
